src/python/pants/build_graph/build_configuration.py

```diff
@@ -186,6 +189,27 @@ def union_rules(self) -> Dict[Type, OrderedSet[Type]]:
         types]."""
         return self._union_rules
 
+    def register_targets(self, targets: Union[typing.Iterable[Type[Target]], Any]) -> None:
+        """Registers the given target types."""
+        if not isinstance(targets, Iterable):
+            raise TypeError(
+                f"The entrypoint `targets` must return an iterable. Given {repr(targets)}"
+            )
+        bad_elements = [
+            tgt_type
+            for tgt_type in targets
+            if not isinstance(tgt_type, type) or not issubclass(tgt_type, Target)
+        ]
+        if bad_elements:
+            raise TypeError(
+                "Every element of the entrypoint `targets` must be a subclass of "
+                f"{Target.__name__}. Bad elements: {bad_elements}."
+            )
+        self._targets = FrozenOrderedSet(sorted(targets, key=lambda target_type: target_type.alias))
```

Comment:
Sorting is for determinism with the engine. 

I wasn't sure where to put this sorting. I considered putting it in `RegisteredTargetTypes.create()`. But I went with here because `BuildCOnfiguration` does appear to get passed around in the engine, and this is an attribute on that type, so it's better to sort as early as possible.

Reply:
Assuming that the set maintains order, the input order should be "ok": effectively, the order that `plugins`/`backends` are loaded in.

But note that these register methods are called once per backend: you will want to extend your ordered set rather than re-creating it each time.

Reply:
Good catch! Thank you!